File: src/opserver/redis_sentinel_client.py
#
# Copyright (c) 2013 Juniper Networks, Inc. All rights reserved.
#

import redis
import gevent

#
# Redis Sentinel Client
#


class RedisSentinelClient(object):
    def __init__(self, sentinel_info, services, master_change_event_handler,
                 logger):
        self._sentinel_info = sentinel_info
        self._services = services
        self._master_change_event_handler = master_change_event_handler
        self._logger = logger
        self._sentinel = None
        self._redis_masters = {}
        self._connect()
        gevent.spawn(self.sentinel_listener)
    #end __init__

    def _connect(self):
        try:
            self._sentinel = redis.StrictRedis(self._sentinel_info[0],
                                               self._sentinel_info[1])
            self._sentinel.execute_command('PING')
        except redis.exceptions.ConnectionError:
            self._sentinel = None
    #end _connect

    def _execute_sentinel_command(self, *args, **kwargs):
        if self._sentinel:
            try:
                return self._sentinel.execute_command('SENTINEL', *args,
                                                      **kwargs)
            except redis.exceptions.ConnectionError:
                return -1
        return None
    #end _execute_sentinel_command

    def _get_redis_master_from_sentinel(self, service):
        try:
            host, port =\
                self._execute_sentinel_command('get-master-addr-by-name',
                                               service)
            redis_master = (host, int(port))
        except Exception as e:
            redis_master = None
            self._logger.error(
                'Failed to get redis master for service "%s" from sentinel'
                % (service))
        finally:
            self._logger.info('Redis master for service %s is %s:%d'
                              % (service, redis_master[0], redis_master[1]))
            self._update_redis_master(service, redis_master)
            return redis_master
    #end _get_redis_master_from_sentinel

    def _get_redis_masters_from_sentinel(self):
        for service in self._services:
            self._get_redis_master_from_sentinel(service)
    #end _get_redis_masters_from_sentinel

    def _update_redis_master(self, service, redis_master):
        try:
            old_redis_master = self._redis_masters[service]
        except KeyError:
            old_redis_master = None
        finally:
            if old_redis_master != redis_master:
                if old_redis_master is not None:
                    del self._redis_masters[service]
                if redis_master is not None:
                    self._redis_masters[service] = redis_master
                    self._logger.info(
                        'Update Redis master %s:%d for service "%s"'
                        % (redis_master[0], redis_master[1], service))
                self._master_change_event_handler(service, redis_master)
            else:
                self._logger.info('No change in Redis master for service %s'
                                  % (service))
    #end _update_redis_master

    # public functions

    def get_redis_master(self, service):
        try:
            redis_master = self._redis_masters[service]
        except KeyError:
            return None
        else:
            return redis_master
    #end get_redis_master

    def sentinel_listener(self):
        while True:
            while self._sentinel is None:
                gevent.sleep(2)
                self._connect()
            self._logger.info('Connected to sentinel %s:%d'
                              % (self._sentinel_info[0],
                                 self._sentinel_info[1]))
            # update redis masters and trigger callback, if there is
            # a change in redis mastership for any service
            self._get_redis_masters_from_sentinel()
            pubsub = self._sentinel.pubsub()
            pubsub.psubscribe('*')
            while True:
                try:
                    msg = pubsub.listen().next()
                except Exception as err:
                    self._logger.error('Error reading message from sentinel')
                    self._sentinel = None
                    break
                else:
                    self._logger.debug('Sentinel message: %s' % (str(msg)))
                    if (msg['channel'] == '-sdown' or
                            msg['channel'] == '-odown'):
                        data = msg['data'].split()
                        if len(data) == 0:
                            self._logger.error(
                                'Failed to decode sentinel message')
                            continue
                        # Ignore state change of non-master
                        if data[0] == 'master':
                            # the format of channels "-sdown" and "-odown" for
                            # redis master is:
                            # master <master-name> <ip> <port>
                            if len(data) != 4:
                                self._logger.error(
                                    'Failed to decode sentinel message')
                                continue
                            self._logger.info(
                                'Redis master up for service %s [%s:%s]'
                                % (data[1], data[2], data[3]))
                            if data[1] in self._services:
                                redis_master = (data[2], int(data[3]))
                                self._update_redis_master(data[1],
                                                          redis_master)
                    # the channel 'switch-master' and 'redirect-to-master' is
                    # of the format:
                    # <master-name> <oldip> <old-port> <new-ip> <new-port>
                    elif (msg['channel'] == '+switch-master' or
                            msg['channel'] == '+redirect-to-master'):
                        data = msg['data'].split()
                        if len(data) != 5:
                            self._logger.error(
                                'Failed to decode sentinel message')
                            continue
                        self._logger.info(
                            'Redis master change for service %s' +
                            '[%s:%s => %s:%s]'
                            % (data[0], data[1], data[2], data[3], data[4]))
                        if data[0] in self._services:
                            redis_master = (data[3], int(data[4]))
                            self._update_redis_master(data[0], redis_master)
    #end sentinel_listerner

#end RedisSentinelClient

File: src/analytics/redis_sentinel_client.cc
/*
 * Copyright (c) 2013 Juniper Networks, Inc. All rights reserved.
 */

#include "base/logging.h"
#include "redis_sentinel_client.h"

RedisSentinelClient::RedisSentinelClient(
                        EventManager *evm, 
                        const std::string& sentinel_ip,
                        unsigned short sentinel_port,
                        RedisServices& services,
                        RedisMasterUpdateCb redis_master_update_cb) : 
        evm_(evm), 
        sentinel_ip_(sentinel_ip), 
        services_(services),
        redis_master_update_cb_(redis_master_update_cb),
        rx_msg_count_(0),
        rx_msg_cnt_since_last_conn_(0),
        sentinel_msg_cb_(NULL),
        getmaster_timer_(*evm->io_service()) {
    sentinel_conn_.reset(new RedisAsyncConnection(evm, sentinel_ip, sentinel_port, 
            boost::bind(&RedisSentinelClient::ConnUp, this),
            boost::bind(&RedisSentinelClient::ConnDown, this)));
    Connect();
}

RedisSentinelClient::~RedisSentinelClient() {
    boost::system::error_code ec;
    getmaster_timer_.cancel(ec);
}

void RedisSentinelClient::Connect() {
    sentinel_conn_.get()->RAC_Connect();
}

void RedisSentinelClient::ConnUp() {
    LOG(DEBUG, "Redis Sentinel Connection Up");
    evm_->io_service()->post(boost::bind(
                &RedisSentinelClient::ConnUpPostProcess, this));
}

void RedisSentinelClient::ConnUpPostProcess() {
    sentinel_msg_cb_ = boost::bind(&RedisSentinelClient::ProcessSentinelMsg,
                                   this, _1, _2, _3);
    sentinel_conn_.get()->SetClientAsyncCmdCb(sentinel_msg_cb_);
    GetRedisMasters();
    sentinel_conn_.get()->RedisAsyncCommand(NULL, "PSUBSCRIBE *");
}

void RedisSentinelClient::ConnDown() {
    LOG(DEBUG, "Redis Sentinel Connection Down");
    boost::system::error_code ec;
    getmaster_timer_.cancel(ec);
    rx_msg_cnt_since_last_conn_ = 0;
    evm_->io_service()->post(boost::bind(&RedisSentinelClient::Connect, this));
}

void RedisSentinelClient::GetRedisMasters() {
    RedisServices::const_iterator it;
    for (it = services_.begin(); it != services_.end(); ++it) {
        sentinel_conn_.get()->RedisAsyncCommand(NULL, 
                    "SENTINEL get-master-addr-by-name %s", (*it).c_str());
    }
}

void RedisSentinelClient::UpdateRedisMaster(const std::string& service,
                                            const std::string& redis_ip,
                                            unsigned short redis_port) {
    RedisMasterInfo redis_master(redis_ip, redis_port);
    RedisMasterMap::iterator it = redis_master_map_.find(service);
    if (it != redis_master_map_.end()) {
        if (it->second == redis_master) {
            LOG(DEBUG, "No change in Redis master for service " << service);
            return;
        } else {
            redis_master_map_.erase(it);
        }
    }
    LOG(DEBUG, "Update Redis master " << redis_ip << ":" << redis_port <<
        " for service " << service);
    redis_master_map_.insert(std::make_pair(service, redis_master));
    redis_master_update_cb_(service, redis_ip, redis_port);
}

void RedisSentinelClient::ProcessSentinelMsg(const redisAsyncContext *c, 
                                             void *msg, void *privdata) {
    redisReply *sentinel_msg = (redisReply*)msg;

    LOG(DEBUG, "Received message from Redis Sentinel");

    if (!sentinel_msg) {
        LOG(DEBUG, "NULL Sentinel message");
        return;
    }

    rx_msg_count_++;
    rx_msg_cnt_since_last_conn_++;

    if (sentinel_msg->type == REDIS_REPLY_ARRAY) {
        LOG(DEBUG, "Redis Sentinel Message: " << sentinel_msg->elements);
        for (int i = 0; i < (int)sentinel_msg->elements; i++) {
            if (sentinel_msg->element[i]->type == REDIS_REPLY_STRING) {
                LOG(DEBUG, "Element" << i << " == " << 
                    sentinel_msg->element[i]->str);
            } else {
                LOG(DEBUG, "Element" << i << " type == " << 
                    sentinel_msg->element[i]->type);
            }
        }
    } else {
        LOG(DEBUG, "Redis Sentinel message type == " << sentinel_msg->type);
        boost::system::error_code ec;
        getmaster_timer_.expires_from_now(boost::posix_time::seconds(RedisSentinelClient::kGetMasterTime), ec);
        getmaster_timer_.async_wait(boost::bind(&RedisSentinelClient::GetRedisMasters, this));
        return;
    }

    if (!strncmp(sentinel_msg->element[0]->str, "psubscribe", 
                 strlen("psubscribe"))) {
        // Ignore response to PSUBSCRIBE *
        return;
    }

    if (!strncmp(sentinel_msg->element[0]->str, "pmessage", 
                 strlen("pmessage"))) {
        if (sentinel_msg->elements != 4) {
            LOG(ERROR, "Unknown Sentinel message");
            return;
        }
        if (!strncmp(sentinel_msg->element[2]->str,
                     "-sdown", strlen("-sdown")) ||
            (!strncmp(sentinel_msg->element[2]->str,
                      "-odown", strlen("-odown")))) {
            std::vector<std::string> channel_msg;
            boost::split(channel_msg, sentinel_msg->element[3]->str,
                         boost::is_any_of(" "), boost::token_compress_on);
            if (!channel_msg.size()) {
                LOG(ERROR, "Failed to decode message " << 
                    sentinel_msg->element[2]->str << "from sentinel");
                return;
            }
            if (strncmp(channel_msg[0].c_str(), "master", strlen("master"))) {
                // Ignore state change of non-master
                return;
            }
            /*
             * the format of channels "-sdown" and "-odown" for redis master is
             * master <master-name> <ip> <port> 
             */
            if (channel_msg.size() != 4) {
                LOG(ERROR, "Failed to decode message " << 
                    sentinel_msg->element[2]->str << "from sentinel");
                return;
            }
            RedisServices::const_iterator it = 
                std::find(services_.begin(), services_.end(), 
                          channel_msg[1]);
            if (it == services_.end()) {
                LOG(DEBUG, "Received Redis Sentinel Message for unknown master " <<
                    channel_msg[1]);
                return;
            }
            LOG(INFO, "Redis Master Up for service " << channel_msg[1] <<
                " [" << channel_msg[2] << ":" << channel_msg[3] << "]");
            UpdateRedisMaster(*it, channel_msg[2], 
                              atoi(channel_msg[3].c_str()));
        } else if (!strncmp(sentinel_msg->element[2]->str,
                            "+switch-master", strlen("+switch-master")) ||
                   (!strncmp(sentinel_msg->element[2]->str, 
                             "+redirect-to-master",
                             strlen("+redirect-to-master")))) {
            /* 
             * the channel 'switch-master' and 'redirect-to-master' is 
             * of the format:
             *
             * <master-name> <old-ip> <old-port> <new-ip> <new-port>
             */
            std::vector<std::string> redis_master_info;
            boost::split(redis_master_info, sentinel_msg->element[3]->str,
                         boost::is_any_of(" "), boost::token_compress_on);
            if (redis_master_info.size() != 5) {
                LOG(ERROR, "Failed to decode Redis Sentinel message " <<
                    sentinel_msg->element[2]->str);
                return;
            }
            RedisServices::const_iterator it = 
                std::find(services_.begin(), services_.end(), 
                          redis_master_info[0]);
            if (it == services_.end()) {
                LOG(DEBUG, "Received Redis Sentinel Message for unknown master " <<
                    redis_master_info[0]);
                return;
            }
            LOG(INFO, "Redis Master change for service " << 
                redis_master_info[0] << " [" << redis_master_info[1] << ":" <<
                redis_master_info[2] << "] => [" << redis_master_info[3] <<
                ":" << redis_master_info[4] << "]");
            UpdateRedisMaster(*it, redis_master_info[3],
                              atoi(redis_master_info[4].c_str()));
        } 
    } else {
        // this is response to "SENTINEL get-master-addr-by-name"
        assert(rx_msg_cnt_since_last_conn_ <= services_.size());
        if (sentinel_msg->elements != 2) {
            LOG(ERROR, "Unknown Redis Sentinel message");
            return;
        }
        LOG(INFO, "Redis Master for service " << 
            services_[rx_msg_cnt_since_last_conn_-1] << " is " << 
            sentinel_msg->element[0]->str << ":" <<
            atoi(sentinel_msg->element[1]->str));
        UpdateRedisMaster(services_[rx_msg_cnt_since_last_conn_-1],
                          sentinel_msg->element[0]->str,
                          atoi(sentinel_msg->element[1]->str));
    }
}
